ocaml/xapi/certs.ml

```diff
@@ -0,0 +1,414 @@
+
+  let decode_multi pem =
+    Cstruct.of_string pem |> X509.Certificate.decode_pem_multiple
+
+  let read_host path =
+    match Sys.file_exists path with
+    | true ->
+        let* pems =
+          Pem.parse_file_using Pem.host_pem path |> function
+          | Error msg ->
+              R.error_msg msg
+          | Ok _ as ok ->
+              ok
+        in
+        let* private_key = decode_key pems.Pem.private_key in
+        let* cert = decode_cert pems.Pem.host_cert in
+        let* intermediares =
+          decode_multi (String.concat "\n" pems.Pem.other_certs)
+        in
+        Ok (Some {private_key; cert; intermediares})
+    | false ->
+        Ok None
+
+  let read_cr path =
+    let name = Filename.basename path in
+    let* pem =
+      Pem.parse_file_using Pem.cert path |> function
+      | Error msg ->
+          R.error_msg msg
+      | Ok _ as ok ->
+          ok
+    in
+    let* cert = decode_cert pem in
+    Ok {cert; name; deleted= false}
+
+  let read_ca path =
+    let name = Filename.basename path |> Filename.chop_extension in
+    let* uuid =
+      match Uuidm.of_string name with
+      | Some uuid ->
+          Ok uuid
+      | None ->
+          R.error_msgf "%s is not a uuid filename" path
+    in
+    let* pem =
+      Pem.parse_file_using Pem.cert path |> function
+      | Error msg ->
+          R.error_msg msg
+      | Ok _ as ok ->
+          ok
+    in
+    let* cert = decode_cert pem in
+    Ok {cert; uuid; deleted= false}
+
+  (* verrify that all values in a list are [Ok] and otherwise return
+     the first [Error] value *)
+  let all_ok xs =
+    let rec loop acc = function
+      | [] ->
+          Ok (List.rev acc)
+      | Ok x :: xs ->
+          loop (x :: acc) xs
+      | (Error _ as e) :: _ ->
+          e
+    in
+    loop [] xs
+
+  let read_crs_dir path =
+    filter_dir (has_extension "*.pem") path |> List.map read_cr |> all_ok
+
+  let read_ca_dir path =
+    filter_dir (has_extension "*.pem") path |> List.map read_ca |> all_ok
+
+  let init paths =
+    with_mutex @@ fun () ->
+    let* host = read_host paths.host_cert_path in
+    let* cr = read_crs_dir paths.ca_crls_dir in
+    let* ca = read_ca_dir paths.ca_certs_dir in
+    t := Some {host; ca; cr; paths} ;
+    Ok ()
+
+  let write_to path content =
+    let perform () =
+      let fd = Unix.(openfile path [O_WRONLY; O_TRUNC; O_CREAT] 0o400) in
+      defer (fun () -> Unix.close fd) @@ fun () ->
+      ignore @@ Unix.write_substring fd content 0 (String.length content)
+    in
+    R.trap_exn perform () |> R.error_exn_trap_to_msg
+
+  let sync_ca dir (ca : ca) =
+    let name = Printf.sprintf "%s.pem" (Uuidm.to_string ca.uuid) in
+    let path = Filename.concat dir name in
+    let rm path = Sys.remove path in
+    match ca.deleted with
+    | true ->
+        R.trap_exn rm path |> R.error_exn_trap_to_msg
+    | false ->
+        let pem = X509.Certificate.encode_pem ca.cert |> Cstruct.to_string in
+        write_to path pem
```

Comment:
I think we need to sync with the bundle here, which is more complicated :(

Reply:
Oh I see now. Is the idea that we write the CA certs to individual files, and then concatenate them all to the bundle file? (Though I don't see the code for that)